server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java

```diff
@@ -460,6 +456,20 @@ private SubFieldInfo buildPhraseInfo(FieldType fieldType, TextFieldType parent)
         @Override
         public TextFieldMapper build(MapperBuilderContext context) {
             MultiFields multiFields = multiFieldsBuilder.build(this, context);
+
+            // If synthetic source is used we need to either store this field
+            // to recreate the source or use keyword multi-fields for that.
+            // So if there are no suitable multi-fields we will default to
+            // storing the field without requiring users to explicitly set 'store'.
+            //
+            // If 'store' parameter was explicitly provided we'll let it pass and
+            // fail in TextFieldMapper#syntheticFieldLoader later if needed.
+            if (store.isSet() == false
```

Comment:
Ideally we would do default values and validation via the `Parameter` class in the `Builder` constructor in the file. Something like this:

```
this.store = Parameter.storeParam(m -> ((TextFieldMapper) m).store, () -> syntheticSource).addValidator(value ->    {
      if (syntheticSource && value == false) {
         throw new IllegalArgumentException("If synthetic source is enabled, then store cannot be set to false");
      }
   });
```

The issue is that now we can't see the details we care about here in the multi fields (`MultiFields.Builder` only keep a map with a `Function` per field name around). 

Note that this way does require [more plumbing and additional changes](https://gist.github.com/martijnvg/2f2067274a8a1d62b4f6353a2979ef93).

If we like to do this approach we may need to make changes to `MultiFields.Builder`, so that we know whether there is a keyword field (either stored or doesn't have normalizer and ignore above) as multi field.

Reply:
I tried this approach and there is a problem. It is not possible to have `syntheticSource` flag with proposed impl from gist in `MappingParserContext` because it will return if synthetic source is enabled for _currently existing_ mapping or empty mapping in case of index creation. `mappingLookup` will only be updated as a last step after parsing and building new mapping. 

Value that we need is here https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/mapper/MappingParser.java#L147 and it is not available to field mapper builders since builders are created in `RootObjectMapper#parse` which happens before that line.

We could try:
1. Creating metadata field mappers before `RootObjectMapper`. I assume that there is a good reason for that ordering so this is unlikely a solution.
2. Support storing text fields by default only for TSDB indices (and any future index modes that will have synthethic source always enabled) instead of for any index with synthetic source enabled. This works since we know source is synthetic even without parsing the mapping. I think we are mostly interested in this use case anyway so this may be a good solution.
3. Keep initial implementation from this PR.

Reply:
Thanks for writing down this analysis. It is kind of a pain that we know synthetic source is used after the builders are created. I don't think it is worth changing the construction of mapping builders, since it most likely is a tricky change. 

Regarding option 2, we know if index mode has been set to time series that synthetic source is enabled, and this will also be the case for the new logging index mode. Given that index mode comes from index setting, we know at the time mapping builders are created what the index mode has been set to. So we can use that as well for determining the default value of the `store` attribute. So let's do option 2?